graph/src/data/store/scalar.rs

```diff
@@ -10,14 +13,158 @@ use stable_hash::{
 };
 use std::convert::{TryFrom, TryInto};
 use std::fmt::{self, Display, Formatter};
+use std::io::Write;
 use std::ops::{Add, Div, Mul, Rem, Sub};
 use std::str::FromStr;
 
 pub use num_bigint::Sign as BigIntSign;
 
-// Caveat: The exponent is currently an i64 and may overflow.
-// See https://github.com/akubera/bigdecimal-rs/issues/54.
-pub type BigDecimal = bigdecimal::BigDecimal;
+/// All operations on `BigDecimal` return a normalized value.
+// Caveat: The exponent is currently an i64 and may overflow. See
+// https://github.com/akubera/bigdecimal-rs/issues/54.
+// Using `#[serde(from = "BigDecimal"]` makes sure deserialization calls `BigDecimal::new()`.
+#[derive(
+    Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, AsExpression, FromSqlRow,
+)]
+#[serde(from = "bigdecimal::BigDecimal")]
+#[sql_type = "diesel::sql_types::Numeric"]
+pub struct BigDecimal(bigdecimal::BigDecimal);
+
+impl From<bigdecimal::BigDecimal> for BigDecimal {
+    fn from(big_decimal: bigdecimal::BigDecimal) -> Self {
+        BigDecimal(big_decimal).normalize()
+    }
+}
+
+impl BigDecimal {
+    pub fn new(digits: BigInt, exp: i64) -> Self {
+        Self::from(bigdecimal::BigDecimal::new(digits.0, -exp))
+    }
+
+    pub fn as_bigint_and_exponent(&self) -> (num_bigint::BigInt, i64) {
+        self.0.as_bigint_and_exponent()
+    }
+
+    pub(crate) fn digits(&self) -> u64 {
+        self.0.digits()
+    }
+
+    // Copy-pasted from `bigdecimal::BigDecimal::normalize`. We can use the upstream version once it
+    // is included in a released version supported by Diesel.
+    pub fn normalize(&self) -> BigDecimal {
+        let (bigint, exp) = self.0.as_bigint_and_exponent();
+        let (_, mut digits) = bigint.to_u32_digits();
+        let mut digits_iter = digits.iter().rev();
+        let mut trailing_count = 0;
+        while digits_iter.next() == Some(&0) {
+            trailing_count += 1;
+        }
+        digits.truncate(digits.len() - trailing_count as usize);
+        let int_val = num_bigint::BigInt::new(self.0.sign(), digits);
```

Comment:
After bringing in the tests I also added the line `assert_eq!(not_normalized.normalize().to_string(), string);` between the other asserts. This causes the tests to fail:
```
thread 'data::store::scalar::test::test_normalize' panicked at 'assertion failed: `(left == right)`
  left: `"0.10"`,
 right: `"0.1"`', graph/src/data/store/scalar.rs:602:13
```

Reply:
Huge thanks for testing this better than I did! The normalization was totally wrong because I wasn't getting the digits in the correct way, and the test was wrong because it was testing numerical equality and not the equality in the representation. Fixed now, thanks again!

Reply:
Also I'm normalizing `0` now.